examples/clean-state/src/index.js

```diff
@@ -3,17 +3,17 @@ import { NearContract, NearBindgen, call, view, near, LookupMap } from 'near-sdk
 @NearBindgen
 class CleanState extends NearContract {
     @call
-    clean(keys) {
+    clean({keys}) {
```

Comment:
Is it possible to modify each method, and replace them with one that already inlines `{` `}` between all the arguments? We can get rid of forcing the users to constantly write these brackets otherwise and would make the classes look a lot simpler

Reply:
It is possible as we have full access to AST. However I think it's controversial to do so - it's a hidden magic turns the user contract to non standard JavaScript.

Reply:
hmm, but would there be a case where the contract is used as standard javascript though? We don't have any unit tests, and workspaces just makes RPC calls anyways. And the thing is, it's all very magicful anyways when we're sending the arguments across the wire onto the network. Plus if we ever decide to support borsh, the brackets `{}` would make it seem weird in that case

Reply:
During contract development you are getting errors. The first place you go is *contract".js file, because error givese you line where it happened. Long magical code will be unfortunate here.

Reply:
Not sure how it will affect borsh.

Reply:
>  The first place you go is *contract".js file, because error givese you line where it happened. Long magical code will be unfortunate here

aren't we generating a separate contract.js file anyways? So error output (or the line the error occurred) would just match up with the generated file then or am I missing something? The generated file should have these inplaced `{}` too so I don't see an issue there.

> Not sure how it will affect borsh

Not entirely sure how the structure of borsh could like, but the brackets `{}` would match more closely to JSON, while borsh could just be any plain old object that's not necessarily a dictionary. Depends on how we want to serialize/deserialize such an object

Reply:
> Plus if we ever decide to support borsh, the brackets {} would make it seem weird in that case

When you make this kind of signature:
```
function rand({ param1, param2 })
```
argument encoding with borsh, it can't work since the order of param1 and param2 is not guaranteed. You can use a borsh-encoded hashmap which has param names and values, but that violates to you want a dense, succinct benefit from borsh. So practically it has to be:
```
function rand(param1, param2)
```

Therefore we have named parameter for json, and non named parameters for borsh. With Serhii's design, both can be achieved by expressing in standard JS:
```
function rand({ param1, param2 })
function rand(param1, param2)
```

If we modify code generation to auto add `{` `}`, user cannot achieve the later case.

Reply:
Good points.

```
@call
@borsh
function rand(param1, param2) {}
```

seems quite good.

> If we do decide to support such a thing in the future, I think the syntax of {param1, param2, ...} might be unwieldy to specify which parameters have what serialization format, but we can think about that later since we won't stabilize 1.0 quite yet

The problem is that, even with above `@borsh` you cannot specify per paramter serialization - js decorators can only on function/method/class, not parameters.

Also I want to know how useful is per-param serialization, e.g.:
```
fn rand(&self, #[serializer(json)] param1, #[serializer(borsh)] param2) {}
```
What is the use case of it, is that only happen at callback function of a borsh contract cross contract call a json contract?

Reply:
Assume we want per-param serialization, given JS decorators' limitation:  can only on function/method/class, not parameters. we possibly do this:
```
@call
@customSerializer({'param2': 'borsh'})
function rand(param1, param2) {}
```
In this case either `rand(param1, param2)` or ``rand({param1, param2})` should work. So Serhii's approach is still good